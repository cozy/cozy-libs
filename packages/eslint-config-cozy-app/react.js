'use strict'

const basics = require('./basics')

module.exports = {
  plugins: basics.plugins.concat(['react-hooks']),
  extends: basics.extends.concat(['plugin:react/recommended']),
  parser: basics.parser,
  parserOptions: { ecmaFeatures: { jsx: true } },
  env: basics.env,
  settings: { react: { version: 'latest' } },
  rules: Object.assign({}, basics.rules, {
    'react/prop-types': 'off',
    'react-hooks/rules-of-hooks': 'error',
    'react-hooks/exhaustive-deps': 'warn'
  })
}
