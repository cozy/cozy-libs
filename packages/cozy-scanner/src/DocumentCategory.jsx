import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import { Bold } from 'cozy-ui/transpiled/react/Text'
import Icon from 'cozy-ui/transpiled/react/Icon'
import IconStack from 'cozy-ui/transpiled/react/IconStack'
import { Media, Img, Bd } from 'cozy-ui/transpiled/react/Media'
import ActionMenu, {
  ActionMenuItem,
  ActionMenuHeader
} from 'cozy-ui/transpiled/react/ActionMenu'
import palette from 'cozy-ui/transpiled/react/palette'
import GridItem from 'cozy-ui/transpiled/react/Labs/GridItem'

import CategoryGridItem from './CategoryGridItem'

import styles from './stylesheet.css'

/**
 *
 * The goal of this component is to display a
 * category / type of document and also manage
 * its associated ActionMenu since a category has
 * several associated items.
 *
 * If an item from a category is selected, we display
 * its label and the label from the category
 *
 */
class DocumentCategory extends Component {
  state = {
    isMenuDisplayed: false
  }

  openMenu = () => {
    this.setState({ isMenuDisplayed: true })
  }

  closeMenu = () => {
    this.setState({ isMenuDisplayed: false })
  }

  onSelect = item => {
    const { onSelect } = this.props
    if (onSelect) onSelect(item)
  }
  render() {
    const { isMenuDisplayed } = this.state
    const { category, isSelected, selectedItem, items, t } = this.props
    return (
      <>
        <GridItem onClick={this.openMenu}>
          <CategoryGridItem
            isSelected={isSelected}
            icon={category.icon}
            theme={t(`Scan.themes.${category.label}`)}
            label={
              isSelected ? t(`Scan.items.${selectedItem.label}`) : undefined
            }
          />
        </GridItem>

        {isMenuDisplayed && (
          <ActionMenu
            onClose={this.closeMenu}
            autoclose
            popperOptions={{ strategy: 'fixed' }}
          >
            <ActionMenuHeader>
              <Media>
                <Img>
                  <IconStack
                    foreground={
                      <Icon
                        icon={category.icon}
                        color={palette.dodgerBlue}
                        size={'16'}
                        className={classNames(styles['icon-absolute-centered'])}
                      />
                    }
                    background={
                      <Icon
                        icon={'file-duotone'}
                        size={'32'}
                        color={palette.dodgerBlue}
                      />
                    }
                  />
                </Img>
                <Bd className={'u-ml-1'}>
                  <Bold tag="span" ellipsis>
                    {t(`Scan.themes.${category.label}`)}
                  </Bold>
                </Bd>
              </Media>
            </ActionMenuHeader>
            {items.map(item => {
              return (
                <ActionMenuItem
                  onClick={() => {
                    this.onSelect({
                      categoryLabel: category.label,
                      itemId: item.id
                    })
                  }}
                  key={item.id}
                >
                  {t(`Scan.items.${item.label}`)}
                </ActionMenuItem>
              )
            })}
          </ActionMenu>
        )}
      </>
    )
  }
}

DocumentCategory.propTypes = {
  onSelect: PropTypes.func,
  category: PropTypes.object.isRequired,
  isSelected: PropTypes.bool,
  selectedItem: PropTypes.object.isRequired,
  items: PropTypes.array.isRequired,
  t: PropTypes.func.isRequired
}

export default DocumentCategory
