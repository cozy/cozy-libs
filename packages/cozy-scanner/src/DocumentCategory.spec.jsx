import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
jest.mock('cozy-ui/transpiled/react/utils/color', () => ({
  getCssVariableValue: () => '#fff'
}))

// Popper does not work well inside of jest as it heavily relies on DOM APIs (see https://github.com/popperjs/popper-core/issues/478).
jest.mock('@material-ui/core/Popper', () => {
  return ({ children }) => children
})

import DocumentCategory from './DocumentCategory'
import MuiCozyTheme from 'cozy-ui/transpiled/react/MuiCozyTheme'

describe('DocumentCategory', () => {
  it('should match snapshot if selected and icon', async () => {
    const onSelect = jest.fn()
    const { queryByText, getByText, asFragment } = render(
      <MuiCozyTheme>
        <DocumentCategory
          onSelect={onSelect}
          category={{
            icon: 'test',
            label: 'test'
          }}
          isSelected={true}
          selectedItem={{ label: 'test', id: '1' }}
          items={[
            {
              id: '1',
              label: 'Label1'
            },
            {
              id: '2',
              label: 'Label2'
            }
          ]}
          t={text => text}
        />
      </MuiCozyTheme>
    )
    expect(asFragment()).toMatchSnapshot()
    // Click on the item
    fireEvent.click(getByText('Scan.items.test'))
    expect(asFragment()).toMatchSnapshot()
    // Check if the second item is displayed in the ActionMenu
    expect(() => getByText('Scan.items.Label2')).not.toThrow()
    await wait(() => getByText('Scan.items.Label2'))

    fireEvent.click(getByText('Scan.items.Label2'))
    // Menu should not be there anymore
    expect(queryByText('Scan.items.Label2')).toBeNull()
    fireEvent.click(getByText('Scan.items.test'))
    await wait(() => getByText('Scan.items.Label2'))

    fireEvent.click(getByText('Scan.items.Label2'))
    expect(onSelect).toBeCalledWith({
      itemId: '2',
      categoryLabel: 'test'
    })
  })
})
