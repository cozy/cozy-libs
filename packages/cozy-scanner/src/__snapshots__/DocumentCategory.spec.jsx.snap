// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentCategory should match snapshot if selected and icon 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-item-2 MuiGrid-grid-xs-3-32 MuiGrid-grid-sm-2-45 styles__gridItem-container___387sO"
  >
    <div
      class="grid-item u-pb-half u-ph-half u-bxz  u-ellipsis u-c-pointer border-selected grid-item__selected"
    >
      <div
        class="styles__IconStack-wrapper___10dhG"
      >
        <svg
          class="styles__icon___23x3R"
          height="32"
          style="fill: var(--dodgerBlue);"
          width="32"
        >
          <use
            xlink:href="#file-duotone"
          />
        </svg>
        <div
          class="styles__IconStack-foregroundIcon___ZvY-t"
        >
          <svg
            class="styles__icon___23x3R"
            height="16"
            style="fill: var(--dodgerBlue);"
            width="16"
          >
            <use
              xlink:href="#test"
            />
          </svg>
        </div>
      </div>
      <div
        class="u-flex u-flex-column u-ellipsis u-bxz"
      >
        <span
          class="grid-item-theme u-ellipsis"
        >
          Scan.themes.test
        </span>
        <span
          class="grid-item-label u-ellipsis"
        >
          Scan.items.test
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DocumentCategory should match snapshot if selected and icon 2`] = `
<DocumentFragment>
  <div
    class="MuiGrid-item-2 MuiGrid-grid-xs-3-32 MuiGrid-grid-sm-2-45 styles__gridItem-container___387sO"
  >
    <div
      class="grid-item u-pb-half u-ph-half u-bxz  u-ellipsis u-c-pointer border-selected grid-item__selected"
    >
      <div
        class="styles__IconStack-wrapper___10dhG"
      >
        <svg
          class="styles__icon___23x3R"
          height="32"
          style="fill: var(--dodgerBlue);"
          width="32"
        >
          <use
            xlink:href="#file-duotone"
          />
        </svg>
        <div
          class="styles__IconStack-foregroundIcon___ZvY-t"
        >
          <svg
            class="styles__icon___23x3R"
            height="16"
            style="fill: var(--dodgerBlue);"
            width="16"
          >
            <use
              xlink:href="#test"
            />
          </svg>
        </div>
      </div>
      <div
        class="u-flex u-flex-column u-ellipsis u-bxz"
      >
        <span
          class="grid-item-theme u-ellipsis"
        >
          Scan.themes.test
        </span>
        <span
          class="grid-item-label u-ellipsis"
        >
          Scan.items.test
        </span>
      </div>
    </div>
  </div>
  <div>
    <div
      style="position: fixed; left: 0px; top: 0px; z-index: #fff;"
    >
      <div
        class="styles__c-actionmenu___22Fp1 styles__c-actionmenu--inline___1SXZa"
      >
        <div
          class="styles__media___cSJMp styles__media--top___K9w0I styles__c-actionmenu-item___gODqd"
        >
          <div
            class="styles__bd___1Uv-F u-mh-1 u-stack-xs"
          >
            Scan.items.Label1
          </div>
        </div>
        <div
          class="styles__media___cSJMp styles__media--top___K9w0I styles__c-actionmenu-item___gODqd"
        >
          <div
            class="styles__bd___1Uv-F u-mh-1 u-stack-xs"
          >
            Scan.items.Label2
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
